### PR TITLE
Add processor to remove style rule errors from files formatted with Prettier

### PR DESCRIPTION
**Pull request by jsnajdr**
**Prettier formatting processor**
The `prettier-formatting` processor will read the `formattingRulesToIgnore` key from the config, which is an array of rule IDs.
    
If the file has a `@format` pragma, errors from these rules will be filtered out, as some ESlint rules are incompatible with how Prettier does things.
    
Your `.eslines.json` config is supposed to look like this:
```
{
  "rulesNotToDowngrade": ["no-unused-vars"],
  "formattingRulesToIgnore": ["indent"],
  "remote": "origin/master",
  "processors": {
    "default": "lines-modified,prettier-formatting",
    "master": "parsing-errors"
  }
}
```

**Support for processor pipeline**
Define multiple processors in your `.eslines.json` like this:
```
"processors": {
  "default": "lines-modified,prettier-formatting"
}
```
Then all these processors, separated by comma, will be run one after another, feeding the output of the previous one into the next.

### CODE REVIEW

**Comment by nosolosw** [comment]
Jarda, this is a really great contribution. Thanks! :heart: 

I understand that the primary use for this is the integration of [calypso-prettier](https://github.com/Automattic/calypso-prettier) in [Calypso](https://github.com/Automattic/wp-calypso). But I'd be interested in transforming this into a general purpose feature not directly tied to the current use case. I'm also concerned about the conceptual complexity that comes when we mix things like processor composition and `rulesNotToDowngrade`.

Here's a proposal:

* Make processors composable. Make their configuration more evident.
* Convert `rulesNotToDowngrade` into a processor called `enforce`.
* Rename `prettier-formatting` to `ignore-whenpragma`.

The config would be:

    {
        "processors": {
            "default": "lines-modified, ignore-whenpragma, enforce",
            "my/topic-branch-1": "lines-modified, ignore-whenpragma",
            "my/topic-branch-2": "ignore-whenpragma, enforce",
            "master": "parsing-errors"
        },
        "processorsConfig": {
            "lines-modified": {
                "remote": "origin/master"
            },
            "enforce": {
                "rules": "no-unused-vars"
            },
            "ignore-whenpragma": {
                "rules": "indent"
            }
        }
    }
 

This approach would reduce the conceptual complexity (you just need to see how processors are concatenated to understand what is happening, not reviewing the code) and is flexible enough to accommodate multiple use cases.

We can land different PRs for each of the steps above. I'm available to help.

What do you think?

**Comment by nosolosw** [comment]
This PR https://github.com/Automattic/eslines/pull/6 addresses the two first points of the proposal above. 

@jsnajdr would you mind giving it a review, and when is ready we can refactor this one to create the `ignore-whenpragma` processor?

**Comment by jsnajdr** [comment]
Hello Andrés! I like the idea to restructure the config information -- rather than a pile of options and environment variables, have a separate config object for each processor.

I'm a bit confused by the proposed `ignore-whenpragma` processor. How configurable would it be?
 The initial docblock of the file can look like this:
```js
/**
 * @fileoverview Awesome code
 * @author Jarda
 * @flow
 * @format
 */
```
And it's parsed into a `docblock` object like this:
```js
{
  fileoverview: 'Awesome code',
  author: 'Jarda',
  flow: '',
  format: ''
}
```
Is the pragma name also a part of the configuration? Can I configure the processor to look for `flow` pragmas instead? And what about its value?

The `ignore-whenpragma` is an abstraction where I don't see a clear purpose or more than one use case, so it's hard to evaluate. At this point, I'd rather go with something like `ignore-when-format`.

**Comment by nosolosw** [comment]
If I understood this correctly, the core idea for this processor is that if a file has the `@format` pragma, the rules in `formattingRulesToIgnore` will be downgraded to warnings. Is that right?

What I was thinking is that `eslines` should be agnostic as to how the pragma was added (with calypso-prettier, manually, etc). It only needs to know how to recognize files (the pragma) and which rules to ignore.

Either:

    "ignore-whenpragma": {
            "rules": ["indent"]
    }

or 

    "ignore-whenpragma": {
            "pragma": "format"
            "rules": ["indent"]
    }

would work I guess. Having the pragma configurable is a nice addition. We even could recommend using the @eslines-ignore pragma.

Does this make sense?

**Comment by jsnajdr** [comment]
> If I understood this correctly, the core idea for this processor is that if a file has the @format pragma, the rules in formattingRulesToIgnore will be downgraded to warnings. Is that right?

Yes, that's right, except we want to completely remove the errors, not just downgrade them. We don't want to use these rules for `@format`ted files at all. If it was possible to disable them conditionally in `eslintrc`, we would do it.

> What I was thinking is that eslines should be agnostic as to how the pragma was added (with calypso-prettier, manually, etc).

I understand now -- you don't want to mention the name of the tool that added the `@format` in the processor name. That makes sense. We don't need to make the pragma configurable then. I like the name `ignore-when-format` best. I'm not even sure if "pragma" is the right word to describe that thing with `@` :) The `jest-docblock` parser calls it "pragma", the JSDoc docs call it "tag". But now we're already in full bikeshed-the-name mode :)

> We even could recommend using the @eslines-ignore pragma.

The `@format` one seems to be the emerging standard in the JS community, so I'd rather use it if possible.

**Comment by nosolosw** [comment]
Thanks for the context on `format`. Would you mind sharing some links to other places where that is used for future reference and for me to be educated on that matter?

Ah, I see now that messages are being filtered, not downgraded to warnings. Now that we are revisiting naming, how about using something like `filter-when-format` for this one? Other processors that downgrade rules would be called `downgrade-*`.

**Comment by jsnajdr** [comment]
> Thanks for the context on format. Would you mind sharing some links to other places where that is used for future reference and for me to be educated on that matter?

Here is a description of the Facebook process as described by one of Prettier's co-authors who happens to work there: [Facebook Adoption Update](https://github.com/prettier/prettier/releases/tag/1.3.0). It's a process suitable for a gradual upgrade of a large codebase and we are reusing it almost verbatim.

Here is @samouri article about Prettier in Calypso: p4TIVU-6zb-p2

> Now that we are revisiting naming, how about using something like `filter-when-format` for this one? Other processors that downgrade rules would be called `downgrade-*`.

I like the `filter`/`downgrade` convention a lot! ❤️ Great idea 👍

**Comment by nosolosw** [comment]
#6 has landed, so this PR can now be rewritten to the new processors format.

**Comment by jsnajdr** [comment]
@nosolosw @samouri I just updated this branch -- please have a look :) I'll need to test in Calypso and resolve the README.md conflicts -- will do later today.

**Comment by jsnajdr** [comment]
Updated the docs, tested successfully in Calypso and updated the Calypso PR (Automattic/wp-calypso/pull/17211). I believe the whole thing is ready to be reviewed and merged now.

File: src/lib/filter-when-format.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
+const docblock = require( 'jest-docblock' );
+
+function fileHasFormatPragma( file ) {
+	return file.source && 'format' in docblock.parse( file.source );
+}
+
+// If the file has the @format pragma at the beginning, it means it was formatted with
```
Comment (nosolosw) [inline src/lib/filter-when-format.js]: Is it necessary to keep this comment?

File: src/processors/README.md [hunk 2]
```diff
@@ -40,6 +40,20 @@ Example:
 
 No config needed.
 
+### filter-when-format
+
+`filter-when-format` will check if the first docblock comment of the file contains a `@format` tag. That means that the file was formatted with Prettier or a similar tool and that formatting-only rules (like `indent`) can be ignored for this file. The motivation for this processor is that ESLint rules sometimes have a different opinion about formatting than Prettier, so we want to conditionally turn them of and treat Prettier as the authority.
```
Comment (nosolosw) [inline src/processors/README.md]: How about this wording: 

`filter-when-format` will remove any break for the rules indicated if the first file docblock comment contains a `@format` tag, so these rule breaks will be filtered and won't be passed to the next processor or final report.

The original use case for this processor was trying to use a tool other than `ESLint` to format the code. Sometimes, `ESLint` had a different opinion about the final formatting , so we needed to conditionally turn some rules off if the file was already formatted (like the `indent` rule).

File: package.json [hunk 3]
```diff
@@ -24,6 +24,7 @@
     "test-filter-parsing-errors": "./node_modules/.bin/tape tests/test-filter-parsing-errors.js | ./node_modules/.bin/faucet"
```
Comment (nosolosw) [inline package.json]: I'd like to have a `"test-filter-when-format"` command so we can test things separately on development.

File: README.md [hunk 4]
```diff
@@ -40,13 +40,13 @@ If you rather use node-like pipes, check the [eslint-eslines](https://github.com
...
+`eslines` reads its configuration from a file named `.eslines.json` placed in the root of your git repository. Out of the box, it comes with four ways of post-processing an ESLint report - we call them *processors*: `downgrade-unmodified-lines`, `filter-parsing-errors`, `filter-when-format`, `enforce`.
 
 Example config:
 
 	{
 		"branches": {
-			"default": ["downgrade-unmodified-lines", "enforce"],
+			"default": ["downgrade-unmodified-lines", "filter-when-format", "enforce"],
 			"master": ["filter-parsing-errors"],
 			"my/topic-branch": ["filter-parsing-errors"]
```
Comment (nosolosw) [inline README.md]: Let's use the `filter-when-format` for the `my/topic-branch` and use `downgrade-unmodified-lines` and `enforce` as `default`. I'd like to have those together to clearly explain the difference between `enforce` and not downgrading, but adding a third processor to the mix makes thing a bit more complicated to understand for a first-time user.

**Comment by nosolosw** [comment]
The code looks good and this is ready to go!

I've left some comments on the wording for explaining the processor, config examples, or adding a new test command. Feel free to add it in this PR or I can do it later, before the release.

I can prepare a new release including this processor hopefully on Thursday.

File: src/lib/filter-when-format.js (hunk 1 above)
Comment (jsnajdr) [inline src/lib/filter-when-format.js]: I removed it -- the processor is documented elsewhere in more-or-less the same words.

File: src/processors/README.md (hunk 2 above)
Comment (jsnajdr) [inline src/processors/README.md]: Updated the README with your wording. I just changed the past tense to present tense -- I think the rule will be actively used for a very long time before we fully migrate the Calypso codebase. The past tense suggests that it is a legacy rule that we used to need in very distant past. :)

File: package.json (hunk 3 above)
Comment (jsnajdr) [inline package.json]: Good catch! I had no idea that these test scripts were defined in `package.json`.

File: README.md (hunk 4 above)
Comment (jsnajdr) [inline README.md]: Updated the example to use `filter-when-format` in the `my/topic-branch`. `filter-parsing-errors` is now used only on `master`.

File: package.json [hunk 5]
```diff
@@ -22,8 +22,10 @@
     "test-downgrade-unmodified-lines": "./node_modules/.bin/tape tests/test-downgrade-unmodified-lines.js | ./node_modules/.bin/faucet",
     "test-exit-codes": "./node_modules/.bin/tape tests/test-exit-codes.js | ./node_modules/.bin/faucet",
     "test-filter-parsing-errors": "./node_modules/.bin/tape tests/test-filter-parsing-errors.js | ./node_modules/.bin/faucet"
+    "test-filter-when-format": "./node_modules/.bin/tape tests/test-filter-when-format.js | ./node_modules/.bin/faucet",
```
Comment (samouri) [inline package.json]: this comma needs to be removed and the line above it needs a comma

Comment (jsnajdr) [inline package.json]: Good catch! Fixed.
